[typescript-fetch] Support deepObject serialization in query parameters

Dijky opened the pull request:
Fixes #2233.

@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @nicokoenig @topce 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh`, `./bin/security/{LANG}-petstore.sh` and `./bin/openapi3/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Add support for nested objects in query parameters.

Object type query parameters can currently not be properly serialized (see related issue above).  
This PR allows nested object parameters in the `HTTPQuery` type and extends the `querystring` function of the API runtime library to recursively serialize objects with the `deepObject` style as defined in [OpenAPI 3.0](https://swagger.io/docs/specification/serialization/#query).

This method does *not* support arrays of objects, as this would cause an ambiguous serialization due to the use of the `explode` format for arrays.  
E.g. a query string `person[name]=Alice&person[name]=Bob` could be interpreted as an array of person objects with one name property each, or one person object with an array property `name`.

Feedback is highly appreciated.

macjohnny commented on modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache:
```diff
@@ -183,15 +183,19 @@ export function exists(json: any, key: string) {
+export function querystring(params: HTTPQuery, prefix: string = '') {
     return Object.keys(params)
         .map((key) => {
+            const fullKey = prefix + (prefix.length ? `[${key}]` : key);
```
maybe it should be configurable whether to use square brackets `[` `]` or e.g. dots `.`?

Dijky replied:
The format I chose matches the `deepObject` style as it is [specified](https://swagger.io/docs/specification/serialization/#query) in OpenAPI 3.0.  
While I have nothing against other formats, that would IMO be a topic to discuss with the OAI.

AFAIK, this generator currently only supports the `form` style with `explode` enabled and no other combination, so there is a general lack of configurability for query parameters.

Dijky commented:
I just ran the scripts to update the sample files and commited the results.

Some changes in the generated samples don't look like they come from my work, so I'm not quite sure what to do with them. I commited them separately in ae1aab9.

macjohnny commented on modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache:
```diff
@@ -183,15 +183,19 @@ export function exists(json: any, key: string) {
             if (value instanceof Array) {
-                const multiValue = value.join(`&${encodeURIComponent(key)}=`);
-                return `${encodeURIComponent(key)}=${multiValue}`;
+                const multiValue = value.join(`&${encodeURIComponent(fullKey)}=`);
```
```suggestion
                const multiValue = value.map(singleValue => encodeURIComponent(singleValue))
                .join(`&${encodeURIComponent(fullKey)}=`);
```

Dijky commented:
@macjohnny Nice catch! Interestingly, the encoding of array elements is omitted in the current master as well.

I noticed that the weird changes to the sample files are caused by the scripts. The OpenAPI 2.0 and 3.0 scripts generate different samples to the same folders, so the ultimate outcome depends on the order of script execution.  
This has nothing to do with my changes, so I removed the changes and cleaned up the PR.

Dijky commented:
@macjohnny Thanks für approving my other PR.  
Do you have any further comment on this one here?

Also, please remove the WIP label if you can. It was added by auto-labeler and is not correct any more.  
This PR is ready from my end.

macjohnny commented:
cc @wing328

wing328 commented:
Sorry for the delay. Missed the notification.
